added types for steam-user

vanitasboi opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

peterblazejewicz commented on types/steam-user/index.d.ts:
```diff
@@ -0,0 +1,5456 @@
+
+/// <reference types="node" />
+
+type SteamID = import('steamid');
```
You can use this syntax IMO, which would make this more readable (about the intent):
```ts
import type SteamID = require('steamid');
import type ByteBuffer = require('bytebuffer');
```

vanitasboi replied:
👍

vanitasboi replied:
hat to bump the minimum typescript version to 4.2 for this to work ^^

peterblazejewicz commented on types/steam-user/index.d.ts:
```diff
@@ -0,0 +1,5456 @@
+type SteamID = import('steamid');
+type ByteBuffer = import('bytebuffer');
+import EventEmitter = require('events');
+import SteamChatRoomClient = require('./SteamChatRoomClient');
```
I believe this requires rewrite, there is no 'SteamChatRoomClient' import (file) in original,
this shuld read:
```ts
import SteamChatRoomClient = require('./components/chatroom');
```
which implies a file at `components/chatroom.d.ts`

vanitasboi replied:
okay :D

peterblazejewicz commented on types/steam-user/index.d.ts:
```diff
@@ -0,0 +1,5456 @@
+     */
+    logOff(): void;
+
+    // relog
```
```suggestion
```
please remove comments like this. Better to have those empty, otherwise please use a proper comment:
```ts
/**
 * Log off of Steam gracefully.
 */
 ```

vanitasboi replied:
fair point 😀

peterblazejewicz reviewed (commented):
@vanitasboi looks good. Please only fix that file to reflect native package,
use the 'import type' in the other file and that's all.
Those `enums` in JS are just constants, this could be reflected in definition, but that's implementation detail, not relevant for people using this definition.
e.g. this IMO better reflects actual file definition:
```ts
// enums/EUnverse.d.ts
declare const EUniverse: {
    Invalid: 0;
    Public: 1;
    Beta: 2;
    Internal: 3;
    Dev: 4;

    '0': 'Invalid';
    '1': 'Public';
    '2': 'Beta';
    '3': 'Internal';
    '4': 'Dev';
};

export = EUniverse;
```

vanitasboi commented:
@peterblazejewicz I like the idea and I tried to make it work. But somehow I get the error `Namespace SteamUser has no exported member ... `. Could you maybe take a look again? Maybe I also did it completely wrong...

peterblazejewicz commented:
> @peterblazejewicz I like the idea and I tried to make it work. But somehow I get the error `Namespace SteamUser has no exported member ... `. Could you maybe take a look again? Maybe I also did it completely wrong...

Oh, this is supposed to work. This is for example used in `node` HTTP constants (a namespace with list of http constants export from a module). Sry for the hassle. Can you revert this change (as I wrote, everything else looks OK), so this can be revisited later?

vanitasboi commented:
@peterblazejewicz Thank you for taking your time explaining it, I have reverted the changes and added comments with the URL of this PR and a branch with the changes I reverted, so it can be easily revisited later :)

peterblazejewicz commented:
cool! So just briefly, e.g. in `index.d.ts` you can have a helper type func that craetes proper export type based on any input. That way, in `ECurrencyCode` one would import this type (names are exemplary):

```ts
export type ValueTypes<T> = T[keyof T];
```
import type { ValuesTypes } from '../'

declare const ....

declare namesapce ECurrencyCode {
    type ECurrencyCodeType = ValueTypes<typeof ECurrencyCode>;
}
...
only becaus this is CommonJS, one requires this additional markup. In ESM, one could just import type and export newly created type directly (but that's not a case with this module)

peterblazejewicz reviewed (approved):
LGTM!

vanitasboi commented:
ready to merge <3

ruj commented:
A necessary correction must be made to the `Options` interface, it lacks the **dataDirectory** property.
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/03d5034a61b77fd7bb3fae09f26e1ffbbcedc32a/types/steam-user/index.d.ts#L871-L887
This property is present in the original code, see [here](https://github.com/DoctorMcKay/node-steam-user#datadirectory). (in [code](https://github.com/DoctorMcKay/node-steam-user/blob/master/index.js#L76-L89))

joshuajeschek commented:
Hi @ruj , I just pushed the fix to #58877 (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58877/commits/272c56efed97a206e6c21797fba0f20728bca061)
You could drop a review over there, the rest has already been reviewed! :)
